File: 501. Find Mode in Binary Search Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function(root) {
  let max = 0;
  let curr = 0;
  let nums = [];
  let preNode;
  inOrder(root);

  function inOrder(node) {
    if (!node) return;
    inOrder(node.left);
    if (preNode && node.val === preNode.val) {
      curr++;
    } else {
      curr = 1;
    }
    if (curr === max) {
      nums.push(node.val);
    }
    if (curr > max) {
      nums = [node.val];
      max = curr;
    }
    preNode = node;
    inOrder(node.right);
  }
  return nums;
};

File: 501. Find Mode in Binary Search Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findMode(root *TreeNode) []int {
	stack := []*TreeNode{}
	max := 0
	count := 0
	var prev *TreeNode = nil
	result := make([]int, 0)

	for root != nil || len(stack) > 0 {
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}

		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if prev != nil && root.Val == prev.Val {
			count++

			if count > max {
				result = result[:0]
				result = append(result, root.Val)
				max = count
			} else if count == max {
				result = append(result, root.Val)
			}
		} else {
			count = 1

			if count > max {
				max = count
				result = append(result, root.Val)
			} else if count == max {
				result = append(result, root.Val)
			}
		}

		prev = root
		root = root.Right
	}

	return result
}